Create admin user with CLI

ernestWarwas opened the pull request:
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.13 <!-- see the comment below -->                  |
| Bug fix?        | no                                                       |
| New feature?    | yes                                                       |
| BC breaks?      | no                                                       |
| Deprecations?   | no <!-- don't forget to update the UPGRADE-*.md file --> |
| Related tickets | fixes https://github.com/Sylius/Sylius/issues/14172, replaces #14232                      |
| License         | MIT                                                          |

<!--
 - Bug fixes must be submitted against the 1.11 or 1.12 branch
 - Features and deprecations must be submitted against the 1.13 branch
 - Make sure that the correct base branch is set

 To be sure you are not breaking any Backward Compatibilities, check the documentation:
 https://docs.sylius.com/en/latest/book/organization/backward-compatibility-promise.html
-->

<img width="538" alt="image" src="https://user-images.githubusercontent.com/40125720/206679505-fd99acb5-37f7-4005-8884-17472470f76f.png">
<img width="884" alt="image" src="https://user-images.githubusercontent.com/40125720/206679612-4898391b-40ad-4250-bcd2-096dd8ee7ac3.png">

johbuch committed:
feat: create admin user with CLI

johbuch committed:
chore: improve code

johbuch committed:
feat: add choices for locale code and enabled user

johbuch committed:
fix: use service argument in xml for construct

johbuch committed:
refacto: improve code

johbuch committed:
chore: add blank line

johbuch committed:
chore: add blank line and remove space

johbuch committed:
refacto: switch container to interface in construct

johbuch committed:
refacto: switch foreach to array_map

ernestWarwas committed:
Test for create admin user command

Test for create admin user command

Add another scenario

vvasiloi commented on src/Sylius/Bundle/AdminBundle/Command/CreateAdminUserCommand.php:
```diff
@@ -0,0 +1,104 @@
+    name: 'sylius:admin-user:create',
+    description: 'Create a new admin user'
+)]
+class CreateAdminUserCommand extends Command
```
Is it not final on purpose?

vvasiloi commented on src/Sylius/Bundle/AdminBundle/Command/CreateAdminUserCommand.php:
```diff
@@ -0,0 +1,104 @@
+)]
+class CreateAdminUserCommand extends Command
+{
+    private SymfonyStyle $io;
```
If the class is not final on purpose, then this property should be protected instead of private.

vvasiloi commented on src/Sylius/Bundle/AdminBundle/Command/CreateAdminUserCommand.php:
```diff
@@ -0,0 +1,104 @@
+            return Command::INVALID;
+        }
+
+        do {
+            $email = $this->io->ask('Email', null);
+        }
+        while (!$this->validateEmail($email));
```
Is this according to the code style rules? Shouldn't the `while` be on the same line as the closing braket?

vvasiloi replied:
Maybe create a Question and use `->setValidator` and `setMaxAttempts` instead of looping indefinetely?
https://symfony.com/doc/current/components/console/helpers/questionhelper.html#validating-the-answer

vvasiloi commented on src/Sylius/Bundle/AdminBundle/Command/CreateAdminUserCommand.php:
```diff
@@ -0,0 +1,104 @@
+        }
+
+        do {
+            $email = $this->io->ask('Email', null);
```
How does this work in a non-interactive mode?

ernestWarwas replied:
Ok, so IMO it should not be executable in non-interactive mode and I added a test for it.

vvasiloi replied:
Why not? In some cases it could be automated.

vvasiloi commented on src/Sylius/Bundle/AdminBundle/Command/CreateAdminUserCommand.php:
```diff
@@ -0,0 +1,104 @@
+            return COMMAND::FAILURE;
+        }
+
+        $userName = $this->io->ask('Username', null);
+        $firstName = $this->io->ask('Firstname', null);
+        $lastName = $this->io->ask('Lastname', null);
+        $password = $this->io->askHidden('Password');
```
Again, how does this work in a non-interactive mode?

vvasiloi commented on src/Sylius/Bundle/AdminBundle/Command/CreateAdminUserCommand.php:
```diff
@@ -0,0 +1,104 @@
+        $adminUser->setFirstName($firstName);
+        $adminUser->setLastName($lastName);
+
+        $locales = array_map(static fn (Locale $locale) => $locale->getCode(), $this->localeRepository->findAll());
```
Normally, the admin can choose any locale, not just those enabled for the shop.

ernestWarwas committed:
Email validation moved to question

Email validation moved to question

Email validation moved to question

fix static analyse

Review fixes

Review fixes

build fix

coldic3 commented on src/Sylius/Bundle/AdminBundle/Command/CreateAdminUserCommand.php:
```diff
@@ -0,0 +1,112 @@
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        if (!$input->isInteractive()) {
+            $this->io->error('This command should be run interactively.');
```
```suggestion
            $this->io->error('This command must be run interactively.');
```

coldic3 commented on src/Sylius/Bundle/AdminBundle/Command/CreateAdminUserCommand.php:
```diff
@@ -0,0 +1,112 @@
+
+        $this->io->title('Admin user creation');
+
+        $confirm = $this->io->confirm('Do you want to create an admin user ?', true);
```
```suggestion
        $confirm = $this->io->confirm('Do you want to create an admin user?', true);
```

coldic3 commented on src/Sylius/Bundle/AdminBundle/Command/CreateAdminUserCommand.php:
```diff
@@ -0,0 +1,112 @@
+        }
+
+        $userName = $this->io->ask('Username');
+        $firstName = $this->io->ask('Firstname');
+        $lastName = $this->io->ask('Lastname');
```
```suggestion
        $firstName = $this->io->ask('First name');
        $lastName = $this->io->ask('Last name');
```

crydotsnake commented:
Thanks for working on this! if i can help in anyway, just ping me

GSadee commented on src/Sylius/Bundle/AdminBundle/Command/CreateAdminUserCommand.php:
```diff
@@ -0,0 +1,112 @@
+        $localeCode = $this->io->choice('Select the locale code', $locales, 'en_US');
+        $adminUser->setLocaleCode($localeCode);
+
+        $enabled = $this->io->confirm('Do you want to enabled this admin user?', true);
```
```suggestion
        $enabled = $this->io->confirm('Do you want to enable this admin user?', true);
```

Nek- commented on src/Sylius/Bundle/AdminBundle/Command/CreateAdminUserCommand.php:
```diff
@@ -0,0 +1,112 @@
+
+        $confirm = $this->io->confirm('Do you want to create an admin user ?', true);
+
+        if (!$confirm) {
```
This confirm step seems useless, the user started an admin user creation, he obviously wants it.

Rafikooo replied:
IMO it's ok. There are many CLI commands that ask for confirmation, such as running migrations

Nek- replied:
The point is that you can CTRL+C at first question here, while if the migration command do not ask it runs migrations, what could be dangerous.

Just saying, I actually do not care that much.

vvasiloi replied:
How about adding this confirmation at the end, after all data has been entered, to allow the user to review it? 🤔

Nek- replied:
This would however be a clever feature indeed. (but is a new feature)

Rafikooo replied:
@Nek- , I've removed the confirmation step, it was indeed impractical.
@vvasiloi , I've implemented your idea, the result is attached in the PR description :)

vvasiloi replied:
@Rafikooo looks great! Thank you!

GSadee commented on src/Sylius/Bundle/AdminBundle/Command/CreateAdminUserCommand.php:
```diff
@@ -0,0 +1,112 @@
+
+    private function checkIfAdminUserExists(string $email): bool
+    {
+        return null !== $this->adminUserRepository->findOneByEmail($email);
```
We should check it by canonicalized email

Rafikooo replied:
I think this is fine as the custom `findOneByEmail` method actually checks by the canonical email

GSadee replied:
Yes, this method checks by field `emailCanonical`, which is correct, but the email value passed to this repository method is not canonicalized, so there is a possibility that this check will not work properly.

GSadee commented on src/Sylius/Bundle/AdminBundle/Tests/Command/CreateAdminUserCommandTest.php:
```diff
@@ -0,0 +1,95 @@
+
+final class CreateAdminUserCommandTest extends TestCase
+{
+    /**
+     * @test
+     */
```
```suggestion
    /** @test */
```

Rafikooo committed:
[Command] CreateAdminUserCommand - codestyle improvements

Rafikooo committed:
[Command] CreateAdminUserCommand - invalid email handling

Rafikooo committed:
[Command] CreateAdminUserCommand - summary & creation confirmation

Rafikooo committed:
[Command] CreateAdminUserCommand - code refactor

vvasiloi commented on src/Sylius/Bundle/AdminBundle/Command/CreateAdminUserCommand.php:
```diff
@@ -0,0 +1,147 @@
+        if ($this->adminUserExists($this->canonicalizer->canonicalize($email))) {
+            $this->io->error(sprintf('Admin user with email address %s already exists.', $email));
+
+            return COMMAND::INVALID;
```
```suggestion
            return Command::INVALID;
```

NoResponseMate commented on src/Sylius/Bundle/AdminBundle/Tests/Command/CreateAdminUserCommandTest.php:
```diff
@@ -0,0 +1,156 @@
+            ->willReturn(new AdminUser())
+        ;
+
+        $this->command->setInputs(['Email' => 'SYLius@exaMPLE.com']);
```
```suggestion
        $this->command->setInputs(['email' => 'SYLius@exaMPLE.com']);
```
For consistency's sake

NoResponseMate commented on src/Sylius/Bundle/AdminBundle/Tests/Command/CreateAdminUserCommandTest.php:
```diff
@@ -0,0 +1,156 @@
+    /** @test */
+    public function it_throws_an_exception_if_provided_email_is_not_valid(): void
+    {
+        $this->command->setInputs(['Email' => 'invalid-email']);
```
```suggestion
        $this->command->setInputs(['email' => 'invalid-email']);
```

jakubtobiasz commented on src/Sylius/Bundle/AdminBundle/Command/CreateAdminUserCommand.php:
```diff
@@ -0,0 +1,150 @@
+use Symfony\Component\Console\Style\SymfonyStyle;
+use Symfony\Component\Intl\Locales;
+
+#[AsCommand(
+    name: 'sylius:admin-user:create',
+    description: 'Create a new admin user'
+)]
```
Why do we need it?

GSadee commented on src/Sylius/Bundle/AdminBundle/Command/CreateAdminUserCommand.php:
```diff
@@ -0,0 +1,150 @@
+
+    private function setAdminUserData(AdminUserInterface $adminUser, string $email): void
+    {
+        $userName = $this->io->ask('Username');
```
```suggestion
        $username = $this->io->ask('Username');
```

GSadee commented:
Thank you, Ernest! :tada:
